src/transformStateWithClones.js

```diff
@@ -8,6 +8,32 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let previousState = { ...state };
+  const history = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(previousState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const remove of action.keysToRemove) {
+          delete previousState[remove];
```

Comment:
Similarly, instead of mutating 'previousState' by deleting properties, create a new copy of the state without the specified keys. This ensures that the original object is not mutated, which is a key requirement of the task.